packages/alfa-css/src/value/math-expression/math.ts

```diff
@@ -113,34 +112,61 @@ export class Math<
   public resolve(
     this: Math<"length">,
     resolver: Expression.LengthResolver
-  ): Option<Length<"px">>;
+  ): Result<Length<"px">, string>;
 
   public resolve(
     this: Math<"length-percentage">,
     resolver: Expression.Resolver<"px", Length<"px">>
-  ): Option<Length<"px">>;
+  ): Result<Length<"px">, string>;
 
-  public resolve(
-    this: Math<"number">,
-    resolver: Expression.PercentageResolver
-  ): Option<Number>;
+  public resolve(this: Math<"number">): Result<Number, string>;
 
   public resolve(
     this: Math,
-    resolver: Expression.Resolver<"px", Length<"px">>
-  ): Option<Numeric> {
+    resolver?:
+      | Expression.LengthResolver
+      | Expression.Resolver<"px", Length<"px">>
+  ): Result<Numeric, string> {
     // Since the expressions can theoretically contain arbitrarily units in them,
     // e.g. calc(1px * (3 deg / 1 rad)) is a length (even though in practice
     // they seem to be more restricted), we can't easily type Expression itself
     // (other than with its Kind).
-    const expression = this._expression.reduce(resolver);
-
-    return this.isDimensionPercentage("length")
-      ? // length are also length-percentage, so this catches both.
-        expression.toLength()
-      : this.isNumber()
-      ? expression.toNumber()
-      : None;
+    try {
+      const expression = this._expression.reduce({
+        length: () => {
+          throw new Error(`Missing length resolver for ${this}`);
+        },
+        percentage: () => {
+          throw new Error(`Missing percentage resolver for ${this}`);
+        },
```

Comment:
This is not super clean. It might be better to push the optionality of the resolver further down and return `Result`.
OTOH, `Expression#reduce` should just expect a correct reducer and do its magic…
Anyway, this is hopefully good enough for now and we can revisit it later if neeed.